allennlp/modules/seq2seq_encoders/bidirectional_transformer_encoder.py

```diff
@@ -0,0 +1,265 @@
+        super().__init__()
+        self.w_1 = torch.nn.Linear(input_dim, ff_dim)
+        self.w_2 = torch.nn.Linear(ff_dim, input_dim)
+        self.dropout = torch.nn.Dropout(dropout)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # pylint: disable=arguments-differ
+        return self.w_2(self.dropout(F.relu(self.w_1(x))))
+
+
+class TransformerEncoder(torch.nn.Module):
+    "Core encoder is a stack of N layers"
+    def __init__(self, layer: torch.nn.Module, num_layers: int, return_all_layers: bool = False) -> None:
+        super().__init__()
+        self.layers = util.clone(layer, num_layers)
+        self.norm = LayerNorm(layer.size)
+        self.return_all_layers = return_all_layers
+
+    def forward(self, x, mask):
+        "Pass the input (and mask) through each layer in turn."
+        all_layers = []
+        for layer in self.layers:
+            x = layer(x, mask)
+            if self.return_all_layers:
+                all_layers.append(x)
+
+        if self.return_all_layers:
+            all_layers[-1] = self.norm(all_layers[-1])
+            return all_layers
+        return self.norm(x)
+
+
+class SublayerConnection(torch.nn.Module):
+    """
+    A residual connection followed by a layer norm.
+    Note for code simplicity the norm is first as opposed to last.
+    """
+    def __init__(self, size: int, dropout: float) -> None:
+        super().__init__()
+        self.norm = LayerNorm(size)
+        self.dropout = torch.nn.Dropout(dropout)
+
+    def forward(self, x: torch.Tensor, sublayer: Callable[[torch.Tensor], torch.Tensor]) -> torch.Tensor:
+        "Apply residual connection to any sublayer with the same size."
+        return x + self.dropout(sublayer(self.norm(x)))
+
+
+class EncoderLayer(torch.nn.Module):
+    "Encoder is made up of self-attn and feed forward (defined below)"
+    def __init__(self,
+                 size: int,
+                 self_attn: torch.nn.Module,
+                 feed_forward: torch.nn.Module,
+                 dropout: float) -> None:
+        super().__init__()
+        self.self_attn = self_attn
+        self.feed_forward = feed_forward
+        self.sublayer = util.clone(SublayerConnection(size, dropout), 2)
+        self.size = size
+
+    def forward(self, x: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
+        "Follow Figure 1 (left) for connections."
+        x = self.sublayer[0](x, lambda x: self.self_attn(x, x, x, mask))
+        return self.sublayer[1](x, self.feed_forward)
+
+
+class MultiHeadedAttention(torch.nn.Module):
+    def __init__(self, num_heads: int, input_dim: int, dropout: float = 0.1) -> None:
+        super().__init__()
+        assert input_dim % num_heads == 0, "input_dim must be a multiple of num_heads"
+        # We assume d_v always equals d_k
+        self.d_k = input_dim // num_heads
+        self.num_heads = num_heads
+        self.linears = util.clone(torch.nn.Linear(input_dim, input_dim), 4)
```

Comment:
It's a bit scary to have these stored as a list because fundamentally they're heterogeneous. The first three layers or for projecting the query, key and value respectively and the final layer is for projecting the concatenated heads. Since you're just porting I don't want to request any big changes, but a warning comment here would be great.